contrib/win32/win32compat/fileio.c

```diff
@@ -242,13 +244,48 @@ fileio_pipe(struct w32_io* pio[2])
 	return -1;
 }
 
+static int
+st_mode_to_file_att(int mode, wchar_t * attributes, int buf_size)
+{
+	DWORD att = 0;
+	switch (mode) {
+	case S_IRWXO:
+		swprintf(attributes, buf_size - 1, L"FA");
+		break;
+	case S_IXOTH:
+		swprintf(attributes, buf_size - 1, L"FX");
+		break;
+	case S_IWOTH:
+		swprintf(attributes, buf_size - 1, L"FW");
+		break;
+	case S_IROTH:
+		swprintf(attributes, buf_size - 1, L"FR");
+		break;
+	default:
+		if((mode & S_IROTH) != 0)
+			att |= FILE_GENERIC_READ;
+		if ((mode & S_IWOTH) != 0)
+			att |= FILE_GENERIC_WRITE;
+		if ((mode & S_IXOTH) != 0)
+			att |= FILE_GENERIC_EXECUTE;
+		swprintf(attributes, buf_size - 1, L"%#llx", att);
+		break;		
+	}
+	return 0;
+}
+
 /* maps open() file modes and flags to ones needed by CreateFile */
 static int
-createFile_flags_setup(int flags, int mode, struct createFile_flags* cf_flags)
+createFile_flags_setup(int flags, mode_t mode, struct createFile_flags* cf_flags)
 {
 	/* check flags */
-	int rwflags = flags & 0x3;
-	int c_s_flags = flags & 0xfffffff0;
+	int rwflags = flags & 0x3, c_s_flags = flags & 0xfffffff0, ret = -1;
+	PSECURITY_DESCRIPTOR pSD = NULL;
+	wchar_t sddl[256] = { 0 }, owner_ace[100] = {0}, everyone_ace[100] = {0};
```

Comment:
Put these static limits in MACROS. Add comments to indicate how the buffer limits are derived. Ex. links to msdn detailing these limits.